pkg/endpoint/bpf.go

```diff
@@ -847,7 +847,14 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 		// GH-3897 would fix this by creating a new map to do an atomic swap
 		// with the old one.
 		stats.mapSync.Start()
-		err = e.syncPolicyMap()
+		// If we're doing out first regeneration after a restore, the endpoint is still
+		// referencing the old ipcache at this point in time. Thus, we should only add
+		// new entries in to the policy map without removing stale entries. We will
+		// clean up the stale entries at the end of regeneration.
+		// Once the endpoint is referencing the new ipcache, this is no longer necessary.
+		// See: GH-28645
```

Comment:
Hmm, I can't help but wonder if the better option is to just delete the old policymap and recreate it ~65 lines up at `policymap.OpenOrCreate()`. The existing datapath will continue to operate, and hold the reference to the existing policymap, using its content, meanwhile we unpin that version and create / pin a new policymap which this code can then happily write whatever it wants without any repercussions. The new one would only be plumbed as part of the upcoming BPF setup steps. No new complexity from understanding whether this code has partially calculated or fully synced the desired state into the maps, and no delays to correct enforcement.

Reply:
If we think it's better to continue with the current approach, I would like to see a more comprehensive description of why adding new entries into the map is OK and deleting entries is not OK. I think in principle the idea is that new policymap entries are allowing traffic which should now be allowed, and if we don't delete the old entries for previously-used identities then those will continue to flow. But equally it's possible for the policy calculation to inject new deny rules that now match newly allocated identities and that could cause a subtly different form of the policy drops. More of a corner case, harder to track down.

Furthermore if this doesn't delete the old entries then yes per the commit message it needs to follow up and do that the second time... are we just relying on the periodic policymap sync controller to do that? So if it fails to run or gets delayed or future developers change it, then they could overlook that this extends the period of policy not being quite 100% in sync with the intended policy (eventually consistent, sure, but desynced longer than necessary).

Reply:
IMO creating a new policy map instead of updating the old one as Joe suggests here would be the right thing to do... im my mental model we already did that :-)